File: src/MyPython/PastInterview/Turing/LongestSubStringWithoutRepeatingChar.py
"""
Given a string s, find the length of the longest substring without duplicate characters.

Example 1:

Input: s = "abcabcbb"
Output: 3
Explanation: The answer is "abc", with the length of 3.
Example 2:

Input: s = "bbbbb"
Output: 1
Explanation: The answer is "b", with the length of 1.
Example 3:

Input: s = "pwwkew"
Output: 3
Explanation: The answer is "wke", with the length of 3.
Notice that the answer must be a substring, "pwke" is a subsequence and not a substring.


Constraints:

0 <= s.length <= 5 * 104
s consists of English letters, digits, symbols and spaces.
"""


def longest_substr_without_repeat_char_set(s):
    longest_value=0
    l_index=0
    r_index=0
    result=[]
    observed = set()

    while r_index < len(s):
        c=s[r_index]
        if c in observed:
            observed.remove(c)
            l_index +=1
        else:
            observed.add(c)
            longest_value = max(longest_value, r_index-l_index + 1)
            result=s[l_index:r_index]
        r_index +=1
    return  result,longest_value

def longest_substr_without_repeat_char_array(s):
    longest_value=0
    l_index=0
    r_index=0
    result=[]
    observed = [0]*128  # Assuming ASCII

    while r_index < len(s):
        c=s[r_index]
        l_index = max(observed[ord(c)], l_index)
        longest_value = max(longest_value, r_index-l_index + 1)
        observed[ord(c)] = r_index + 1
        r_index+=1
        result=s[l_index:r_index]
    return  result,longest_value


    # if observed[ord(c)] == 1:
    #         observed[ord(s[l_index])] = 0
    #         l_index +=1
    #     else:
    #         observed[ord(c)] = 1
    #         longest_value = max(longest_value, r_index-l_index + 1)
    #         result=s[l_index:r_index+1]
    #     r_index +=1
    # return  result,longest_value


def run_tests():
    # print(longest_substr_without_repeat_char_set("abcabcbb"))  # Expected: ('abc', 3) or similar
    # print(longest_substr_without_repeat_char_set("bbbbb"))  # Expected: ('b', 1) or similar
    # print(longest_substr_without_repeat_char_set("pwwkew"))  # Expected: ('wke', 3) or similar
    # print(longest_substr_without_repeat_char_set(""))  # Expected: ('', 0)
    # print(longest_substr_without_repeat_char_set("a"))  # Expected: ('a', 1)
    # print(longest_substr_without_repeat_char_set("abcdef"))  # Expected: ('abcdef', 6)

    print(longest_substr_without_repeat_char_array("abcabcbb"))  # Expected: ('abc', 3) or similar
    print(longest_substr_without_repeat_char_array("bbbbb"))  # Expected: ('b', 1) or similar
    print(longest_substr_without_repeat_char_array("pwwkew"))  # Expected: ('wke', 3) or similar
    print(longest_substr_without_repeat_char_array(""))  # Expected: ('', 0)
    print(longest_substr_without_repeat_char_array("a"))  # Expected: ('a', 1)
    print(longest_substr_without_repeat_char_array("abcdef"))  # Expected: ('abcdef', 6)

if __name__ == "__main__":
    run_tests()

File: src/MyJava/PastInverview/Turing/LongestSubStringWithoutRepeatingChar.java
package MyJava.PastInverview.Turing;

/**
 * Given a string, find the length of the longest substring without repeating characters.
 *
 * Examples:
 *
 * Given "abcabcbb", the answer is "abc", which the length is 3.
 * Given "bbbbb", the answer is "b", with the length of 1.
 * Given "pwwkew", the answer is "wke", with the length of 3. Note that the answer must be a substring, "pwke" is a subsequence and not a substring.
 *
 * @author Alestar
 *
 *
 * Explanation:
 *
 * We use an integer array index of size 128 (to cover all ASCII characters) is used to store the last seen position of each character in the string.
 * We maintain a sliding window [i, j] (j = j + 1 in each iteration) and move i to the right side of the last seen position of s[j] if needed.
 * We also update the result if we get a longer substring.
 *
 * Time complexity: O(n)
 * Space complexity: O(1)
 *
 *
 */
public class LongestSubStringWithoutRepeatingChar{
    public int lengthOfLongestSubStrWithoutRepHashSet(String s) {
        int n = s.length();
        int ans = 0;
        java.util.Set<Character> set = new java.util.HashSet<>();
        int i = 0, j = 0;
        while (i < n && j < n) {
            // try to extend the range [i, j]
            if (!set.contains(s.charAt(j))){
                set.add(s.charAt(j++));
                ans = Math.max(ans, j - i);
            }
            else {
                set.remove(s.charAt(i++));
            }
        }
        return ans;
    }
    public int lengthOfLongestSubStrWithoutRepOptimal(String s) {
        int n = s.length();
        int ans = 0;
        //An array index of size 128 (to cover all ASCII characters) is used to store the last seen position of each character in the string.
        int[] index = new int[128]; // current index of character
        // try to extend the range [i, j]
        for (int j = 0, i = 0; j < n; j++) {
            System.out.println("char at : " +  s.charAt(j));
            System.out.println("char index : " +  index[s.charAt(j)]);

            //For each character s[j], the method checks if it has been seen before within the current window ([i, j]).
            //If it has, the start of the window (i) is updated to the maximum of its current value and the position after the last occurrence of s[j] (index[s.charAt(j)]).
            i = Math.max(index[s.charAt(j)], i);
            ans = Math.max(ans, j - i + 1);
            index[s.charAt(j)] = j + 1;
        }
        return ans;
    }

    public static  void main(String[] args) {

        LongestSubStringWithoutRepeatingChar s = new LongestSubStringWithoutRepeatingChar();
        System.out.println("Result (HashSet): " + s.lengthOfLongestSubStrWithoutRepHashSet("abcabcbb"));
        System.out.println("Result (HashSet): " + s.lengthOfLongestSubStrWithoutRepHashSet("bbbb"));
        System.out.println("Result (Optimal): " + s.lengthOfLongestSubStrWithoutRepOptimal("abcabcbb"));
        System.out.println("Result (Optimal): " + s.lengthOfLongestSubStrWithoutRepOptimal("bbbb"));



    }
}
